server/sonar-server/src/main/java/org/sonar/server/computation/step/IntegrateIssuesStep.java

```diff
@@ -0,0 +1,156 @@
+    this.baseIssuesLoader = baseIssuesLoader;
+    this.issueLifecycle = issueLifecycle;
+    this.issueVisitors = issueVisitors;
+  }
+
+  @Override
+  public void execute() {
+    // all the components that had issues before this analysis
+    final Set<String> unprocessedComponentUuids = Sets.newHashSet(baseIssuesLoader.loadUuidsOfComponentsWithOpenIssues());
+
+    new DepthTraversalTypeAwareVisitor(Component.Type.FILE, POST_ORDER) {
+      @Override
+      public void visitAny(Component component) {
+        processIssues(component);
+        unprocessedComponentUuids.remove(component.getUuid());
+      }
+    }.visit(treeRootHolder.getRoot());
+
+    closeIssuesForDeletedComponentUuids(unprocessedComponentUuids);
+  }
+
+  private void processIssues(Component component) {
+    Tracking<DefaultIssue, DefaultIssue> tracking = tracker.track(component);
+    Loggers.get(getClass()).info("-- before: "+ tracking);
+    DiskCache<DefaultIssue>.DiskAppender cacheAppender = issueCache.newAppender();
+    try {
+      issueVisitors.beforeComponent(component, tracking);
+      fillNewOpenIssues(component, tracking, cacheAppender);
+      fillExistingOpenIssues(component, tracking, cacheAppender);
+      closeUnmatchedBaseIssues(component, tracking, cacheAppender);
+      issueVisitors.afterComponent(component);
+    } finally {
+      cacheAppender.close();
+    }
+    Loggers.get(getClass()).info("-- after: "+ tracking);
+  }
+
+  private void fillNewOpenIssues(Component component, Tracking<DefaultIssue, DefaultIssue> tracking, DiskCache<DefaultIssue>.DiskAppender cacheAppender) {
+    Set<DefaultIssue> issues = tracking.getUnmatchedRaws();
+    for (DefaultIssue issue : issues) {
+      issueLifecycle.initNewOpenIssue(issue);
+      process(component, issue, cacheAppender);
+    }
+  }
+
+  private void fillExistingOpenIssues(Component component, Tracking<DefaultIssue, DefaultIssue> tracking, DiskCache<DefaultIssue>.DiskAppender cacheAppender) {
+    for (Map.Entry<DefaultIssue, DefaultIssue> entry : tracking.getMatchedRaws().entrySet()) {
+      DefaultIssue raw = entry.getKey();
+      DefaultIssue base = entry.getValue();
+      issueLifecycle.mergeExistingOpenIssue(raw, base);
+      process(component, raw, cacheAppender);
+    }
+    for (Map.Entry<Integer, DefaultIssue> entry : tracking.getOpenManualIssuesByLine().entries()) {
+      Integer line = entry.getKey();
+      DefaultIssue manualIssue = entry.getValue();
+      issueLifecycle.moveOpenManualIssue(manualIssue, line);
+      process(component, manualIssue, cacheAppender);
+    }
+  }
+
+  private void closeUnmatchedBaseIssues(Component component, Tracking<DefaultIssue, DefaultIssue> tracking, DiskCache<DefaultIssue>.DiskAppender cacheAppender) {
+    for (DefaultIssue issue : tracking.getUnmatchedBases()) {
+      Loggers.get(getClass()).info("--- close base " + issue);
+      // TODO should replace flag "beingClosed" by express call to transition "automaticClose"
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1135)